floatingactionmenu/src/main/java/io/cheesecakelabs/floatingactionmenu/uicomponents/customviews/FloatingActionMenuView.kt

```diff
@@ -0,0 +1,208 @@
+        val floatBtn = view.findViewById<FloatingActionButton>(R.id.menuFloatingButton)
+
+        floatBtn?.run {
+            foreground = context.getContextCompatDrawable(drawableOnMenuCollapsed)
+            setOnClickListener { onFabMenuButtonClicked() }
+        }
+
+        this.addView(view)
+    }
+
+    private fun onFabMenuButtonClicked() {
+        if (!isAnimating) {
+            isAnimating = true
+
+            if (isFabMenuOpened) hideFabMenu()
+            else showFabMenu()
+
+            isFabMenuOpened = !isFabMenuOpened
+        }
+    }
+
+    private fun onMenuItemsSet(floatingMenuItems: MutableList<FloatingMenuItem>) {
+        floatingMenuItems.forEach { floatingMenuItem ->
+            val view = LayoutInflater.from(context).inflate(
+                R.layout.fab_menu_item,
+                this,
+                false
+            )
+            val floatBtn =
+                view.findViewById<FloatingActionButton>(R.id.menuItemFloatingButton)
+
+            floatBtn?.run {
+                foreground = context.getContextCompatDrawable(floatingMenuItem.drawable)
+                floatingButtons.add(this)
+                setOnClickListener { listener?.onMenuItemClicked(floatingMenuItem.position) }
+            }
+
+            this.addView(view)
+        }
+
+        menuFloatingButton.bringToFront()
+        setColorDefault()
+    }
+
+    private fun showFabMenu() {
+        val initialDistance = distanceBetweenItems + menuFloatingButton.height
+
+        floatingButtons.forEachIndexed { index, fab ->
+            fab.visible()
+
+            val distance = initialDistance + (fab.height + distanceBetweenItems) * index
+
+            val viewPropertyAnimator = fab.animate().translationY(-distance)
+
+            if (fab == floatingButtons.last()) {
+                viewPropertyAnimator.setupListener(
+                    onAnimationStart = {
+                        listener?.onMenuExpanded()
+                        menuFloatingButton.foreground =
+                            context.getContextCompatDrawable(drawableOnMenuExpanded)
+                    },
+                    onAnimationEnd = { isAnimating = false }
+                )
+            }
+        }
+    }
+
+    private fun hideFabMenu() {
+        floatingButtons.forEach { fab ->
+            val viewPropertyAnimator = fab.animate().translationY(0f)
+
+            if (fab == floatingButtons.last()) {
+                viewPropertyAnimator.setupListener(
+                    onAnimationStart = {
+                        menuFloatingButton.foreground =
+                            context.getContextCompatDrawable(drawableOnMenuCollapsed)
+
+                    },
+                    onAnimationEnd = {
+                        listener?.onMenuCollapsed()
+                        floatingButtons.forEach { it.gone() }
+                        isAnimating = false
+                    }
+                )
+            }
+        }
+    }
+}
```

Comment:
I think `showFabMenu()` and `hideFabMenu()` also have many things in common that could be moved to a function

Reply:
In this case, I don't see so much similarity between them. I prefer let them separated because they have a lot of things in different order.